### ISSUE TITLE
Move MultiSlit slit attributes to meta within the slit

### ISSUE THREAD

**Issue by jdavies-st** [spacetelescope/jwst#527]
Right now `xstart`, `xsize` and `name` are appended directly to an individual slit in `extract_2d`, while the wcs object is attached to the the slit's `meta`.  This causes problems when repackaging MultiSlit data (say from exposure-based to source-based) with running the `update()` method, which only copies over items from meta.

Currently for MultiSlitModel `model`:

```
model.slits[0].xstart
model.slits[0].xsize
model.slits[0].name
```

Possible change:

```
model.slits[0].meta.xstart
model.slits[0].meta.xsize
model.slits[0].meta.name
```

These values (and the others) should be updated in the schema, so that they get validated when written.  Code upstream would have to be changed too.

Interested in people's thoughts on this.

**Comment by hbushouse** [comment]
I agree that these values should be part of a meta attribute in a slit instance. They start out as part of .meta in the original ImageModel, so they ought to also be within .meta when a MultiSlit is created from that ImageModel. This should be a fairly simple change (I would think) in the MultiSlit schema and also then offers the advantage that they get copied when using the .update method.

**Comment by hbushouse** [comment]
From what I can tell it looks like the MultiSlitModel schema is still setup such that attributes like xstart, xsize, etc. are not contained in a meta object. Is this still an issue to be pursued? If so, it's a big enough change to enough code that we should wait for B7.2.
@nden @jdavies-st

**Comment by nden** [comment]
I am in the process of refactoring this code in #1070 and this will be resolved there. Slip to 7.2.

**Comment by stscicrawford** [comment]
Closing as refactoring occurred in #1070.  If there was more work or that was not the case, please re-open or create a new ticket.

**Comment by jdavies-st** [comment]
Reopening as this work has not been done yet.

**Comment by nden** [comment]
@stscijgbot-jp

**Comment by stscijgbot-jp** [comment]
_This issue is tracked on JIRA as [JP-2175](https://jira.stsci.edu/browse/JP-2175)._

**Comment by stscijgbot-jp** [comment]
_Comment by [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) on [JIRA](https://jira.stsci.edu/browse/JP-2175?focusedCommentId=536531#comment-536531):_

[Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) Just to confirm, this essentially means moving the attributes under slitmeta,

<https://github.com/spacetelescope/jwst/blob/master/jwst/datamodels/schemas/slitmeta.schema.yaml>

under "meta". We need to decide where exactly in "meta".

**Comment by stscijgbot-jp** [comment]
_Comment by [James Davies](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jdavies) on [JIRA](https://jira.stsci.edu/browse/JP-2175?focusedCommentId=536950#comment-536950):_

I'll add that the other "Multi" models also put metadata items directly into the slit list.  So `MultiSpecModel` has

```
for spectrum in model.spec:
    print(spectrum.source_type)
```

which really should be `spectrum.meta.source_type`.

This is true for more schemas:

```
multicombinedspec.schema.yaml
multiextract1d.schema.yaml
multispec.schema.yaml
multiexposure.schema.yaml
multislit.schema.yaml
```

The last two use `slitdata.schema.yaml` which is what we are talking about above in this issue, but it would be good to address these other cases as well, as we've consistently seen bugs of metadata not getting carried from one data product to another.  See JP-2198 and others.

**Comment by stscijgbot-jp** [comment]
_Comment by [Edward Slavich](https://jira.stsci.edu/secure/ViewProfile.jspa?name=eslavich) on [JIRA](https://jira.stsci.edu/browse/JP-2175?focusedCommentId=550053#comment-550053):_

Here's a proposal:

Currently, we have metadata like the following:

```python
MultiCombinedSpecModel.spec[0].source_id
MultiExposureModel.exposures[0].source_id
MultiSlitModel.slits[0].source_id
MultiSpecModel.spec[0].source_id
```

Move source_id and friends to meta.spec within the individual exposure/slit/spec:

```python
MultiCombinedSpecModel.spec[0].meta.spec.source_id
MultiExposureModel.exposures[0].meta.spec.source_id
MultiSlitModel.slits[0].meta.spec.source_id
MultiSpecModel.spec[0].meta.spec.source_id
```

This is a little verbose, but it seems important to namespace this metadata to prevent collisions when merging models together.

**Comment by stscijgbot-jp** [comment]
_Comment by [James Davies](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jdavies) on [JIRA](https://jira.stsci.edu/browse/JP-2175?focusedCommentId=550160#comment-550160):_

In the case of something like `name`, the FITS keyword is SLTNAME, which perhaps implies something like `meta.slit.name`?

Others such as `source_type` map to the SRCTYPE FITS keyword, and that particular one is in the core schema as well under `meta.target.source_type`.

So maybe the items associated with the source itself might be under `target`?  And things associated with the slit under `slit`?  Consistency across the models would clearly be good, and consistency in name with anything in the core schema also good.

As an example, `source_ra` and `source_dec` are actually the same as `meta.target.ra` and `meta.taget.dec`?  Not clear to me.

**Comment by stscijgbot-jp** [comment]
_Comment by [Edward Slavich](https://jira.stsci.edu/secure/ViewProfile.jspa?name=eslavich) on [JIRA](https://jira.stsci.edu/browse/JP-2175?focusedCommentId=550179#comment-550179):_

spacetelescope/stdatamodels#85

**Comment by nden** [comment]
Discussed the implications of these changes with @hbushouse and @eslavich, and decided that this should not be included in build 7.9. It requires changes to data schemas and changes to cal code to point to the new locations of attributes. It's a risk to have these changes in build 7.9. They can be considered for build 8.0.

**Comment by stscijgbot-jp** [comment]
_This issue is tracked on JIRA as [JP-2175](https://jira.stsci.edu/browse/JP-2175)._

**Comment by stscijgbot-jp** [comment]
_Comment by [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) on [JIRA](https://jira.stsci.edu/browse/JP-2175?focusedCommentId=536531#comment-536531):_

[Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) Just to confirm, this essentially means moving the attributes under slitmeta,

<https://github.com/spacetelescope/jwst/blob/master/jwst/datamodels/schemas/slitmeta.schema.yaml>

under "meta". We need to decide where exactly in "meta".

**Comment by stscijgbot-jp** [comment]
_Comment by [James Davies [X]](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jdavies) on [JIRA](https://jira.stsci.edu/browse/JP-2175?focusedCommentId=536950#comment-536950):_

I'll add that the other "Multi" models also put metadata items directly into the slit list.  So `MultiSpecModel` has

```
for spectrum in model.spec:
    print(spectrum.source_type)
```

which really should be `spectrum.meta.source_type`.

This is true for more schemas:

```
multicombinedspec.schema.yaml
multiextract1d.schema.yaml
multispec.schema.yaml
multiexposure.schema.yaml
multislit.schema.yaml
```

The last two use `slitdata.schema.yaml` which is what we are talking about above in this issue, but it would be good to address these other cases as well, as we've consistently seen bugs of metadata not getting carried from one data product to another.  See JP-2198 and others.

**Comment by stscijgbot-jp** [comment]
_Comment by [Edward Slavich](https://jira.stsci.edu/secure/ViewProfile.jspa?name=eslavich) on [JIRA](https://jira.stsci.edu/browse/JP-2175?focusedCommentId=550053#comment-550053):_

Here's a proposal:

Currently, we have metadata like the following:

```python
MultiCombinedSpecModel.spec[0].source_id
MultiExposureModel.exposures[0].source_id
MultiSlitModel.slits[0].source_id
MultiSpecModel.spec[0].source_id
```

Move source_id and friends to meta.spec within the individual exposure/slit/spec:

```python
MultiCombinedSpecModel.spec[0].meta.spec.source_id
MultiExposureModel.exposures[0].meta.spec.source_id
MultiSlitModel.slits[0].meta.spec.source_id
MultiSpecModel.spec[0].meta.spec.source_id
```

This is a little verbose, but it seems important to namespace this metadata to prevent collisions when merging models together.

**Comment by stscijgbot-jp** [comment]
_Comment by [James Davies [X]](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jdavies) on [JIRA](https://jira.stsci.edu/browse/JP-2175?focusedCommentId=550160#comment-550160):_

In the case of something like `name`, the FITS keyword is SLTNAME, which perhaps implies something like `meta.slit.name`?

Others such as `source_type` map to the SRCTYPE FITS keyword, and that particular one is in the core schema as well under `meta.target.source_type`.

So maybe the items associated with the source itself might be under `target`?  And things associated with the slit under `slit`?  Consistency across the models would clearly be good, and consistency in name with anything in the core schema also good.

As an example, `source_ra` and `source_dec` are actually the same as `meta.target.ra` and `meta.taget.dec`?  Not clear to me.

**Comment by stscijgbot-jp** [comment]
_Comment by [Edward Slavich](https://jira.stsci.edu/secure/ViewProfile.jspa?name=eslavich) on [JIRA](https://jira.stsci.edu/browse/JP-2175?focusedCommentId=550179#comment-550179):_

spacetelescope/stdatamodels#86
